app/master/slave.py

```diff
@@ -244,6 +250,87 @@ def update_last_heartbeat_time(self):
     def get_last_heartbeat_time(self) -> datetime:
         return self._last_heartbeat_time
 
+    def _update_slave_registry(self):
+        """
+        Remove shutdown-ed slave from SlaveRegistry.
+        """
+        self._logger.info('Removing slave (url={}; id={}) from Slave Registry.'.format(self.url, self.id))
+        SlaveRegistry.singleton().remove_slave(slave_url=self.url)
+
+
+class SlaveRegistry(Singleton):
+    """
+    SlaveRegistry class is a singleton class which stores and maintains list of connected slaves.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.all_slaves_by_url = {}
```

Comment:
add `# type:` annotations

Reply:
Add type annotations. Also if you are not going address getter changes then make this attribute private.

Reply:
Is there any example of this?

Reply:
`self.all_slaves_by_url = types.DictType.__new__(types.DictType, (), {})` or
`self.all_slaves_by_url = dict()`?

Reply:
https://www.python.org/dev/peps/pep-0484/

Something like:

```
self.all_slaves_by_url = {}  # type: Dict[String,ClusterRunnerSlave]
```